profile-specific-challenges/frontend/components/idioms.jsx

```diff
@@ -48,36 +63,44 @@ export function UseEffectThrashing({ fetchURL, label }) {
 
 export function UseEffectDerivedCalculation() {
   const [remainder, setReminder] = useState();
-  const [clickedTimes, setClickedTimes] = useState();
+  const [clickedTimes, setClickedTimes] = useState(0);
+  const previousClickedTimes = useRef(clickedTimes);
 
   useEffect(() => {
-    setReminder(clickedTimes % 5);
+    if (previousClickedTimes !== clickedTimes) {
+      setReminder(clickedTimes % 5);
+      previousClickedTimes.current = clickedTimes;
+    }
   }, [clickedTimes]);
 
   const handleClick = () => setClickedTimes(clickedTimes + 1);
 
   return (
     <div>
       <button onClick={handleClick}>Add Click Count</button>
-      <span>{sum}</span>
+      <span>{clickedTimes}</span>
       <span>{remainder}</span>
     </div>
   );
 }
 
 export function UseStateDerivedCalculation() {
-  const [remainder, setReminder] = useState();
-  const [clickedTimes, setClickedTimes] = useState();
+  const [remainder, setReminder] = useState(0);
+  const [clickedTimes, setClickedTimes] = useState(0);
+  const previousClickedTimes = useRef(clickedTimes);
 
   const handleClick = () => {
-    setClickedTimes(clickedTimes + 1);
-    setReminder(clickedTimes % 5);
+    if (previousClickedTimes !== clickedTimes) {
+      setClickedTimes(clickedTimes + 1);
+      setReminder((clickedTimes + 1) % 5);
+      previousClickedTimes.current = clickedTimes;
+    }
```

Comment:
Elements that are calculated from existing state can be done on render. Using extra state causes renders, but using ref adds complexity by turning the component stateful.